compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,272 @@
+  rewriter.setInsertionPointToStart(&root.getBody().getBlocks().front());
+
+  // 3. Distribute all operations in the worklist until we reach a fixed
+  // point.
+  LLVM_DEBUG(
+      llvm::dbgs() << "Step 3: Starting distributing collected operations\n");
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    for (unsigned i = 0; i < worklist.size(); ++i) {
+      Operation *op = worklist[i];
+      if (erasedOps.count(op))
+        continue;
+
+      rewriter.setInsertionPoint(op);
+
+      LLVM_DEBUG(llvm::dbgs() << "Trying to distribute: ");
+      LLVM_DEBUG(op->print(llvm::dbgs(), OpPrintingFlags().skipRegions()));
+      LLVM_DEBUG(llvm::dbgs() << "\n");
+
+      if (provider->specializedDistribution(rewriter, op).succeeded()) {
+        LLVM_DEBUG(llvm::dbgs() << "Specialized Distribution Successful\n");
+        changed = true;
+        continue;
+      }
+
+      LogicalResult distributed =
+          TypeSwitch<Operation *, LogicalResult>(op)
+              .Case([&](arith::ConstantOp constantOp) {
+                return distributeConstants(rewriter, constantOp);
+              })
+              .Default([&](Operation *op) { return failure(); });
+
+      // If the operation was distributed, continue with the next one.
+      if (distributed.succeeded()) {
+        LLVM_DEBUG(llvm::dbgs() << "Distribution Successful\n");
+        changed = true;
+        continue;
+      }
+
+      // Try distributing as an elementwise operation.
+      if (OpTrait::hasElementwiseMappableTraits(op)) {
+        if (distributeElementwise(rewriter, op).succeeded()) {
+          LLVM_DEBUG(llvm::dbgs() << "Distribution Successful\n");
+          changed = true;
+        }
+      }
+
+      LLVM_DEBUG(llvm::dbgs() << "Distribution Failed\n");
+    }
+  }
+
+  // 4. Ideally, we should error out here if everything was not distributed.
+  // Currently, I'm not adding it for debugging purposes.
+  // TODO: Add a check here if something was not distributed.
+  LLVM_DEBUG(llvm::dbgs() << "Distribution Finished\n");
+}
+
+LogicalResult
+VectorDistribution::distributeConstants(RewriterBase &rewriter,
+                                        arith::ConstantOp constantOp) {
+  Value constantResult = constantOp.getResult();
+  if (!isa<VectorType>(constantResult.getType()))
+    return failure();
+  auto constant = cast<TypedValue<VectorType>>(constantResult);
+  auto attr = llvm::cast<DenseElementsAttr>(constantOp.getValue());
+
+  // Only handle splat values for now.
+  if (!attr.isSplat())
+    return failure();
+
+  // Replace the original op with the distributed op.
+  Type elementType = constant.getType().getElementType();
+  auto vectorType =
+      VectorType::get(provider->getDistributedShape(constant), elementType);
+  replaceOpWithNewDistributedOp<arith::ConstantOp>(
+      provider, rewriter, constantOp, vectorType,
+      DenseElementsAttr::get(vectorType, attr.getSplatValue<APFloat>()));
```

Comment:
this only handles floats
```suggestion
      SplatElementsAttr::get(vectorType, attr.getSplatValue<Attribute>()));
```